NestCam support and huge overhaul to support official Nest API

technicalpickles opened the pull request:
https://github.com/jkoelker/python-nest/issues/44 is requesting NestCam support, and @tikismoke  pointed out @les69 had extended python-nest with support for it over at https://github.com/les69/calvin-base/blob/c2f59e5701d54e9ffee8c0b6be33874c84cd9a87/calvin/runtime/north/plugins/nest/nest.py

This PR imports those changes, and credits @les69 as the committer. I haven't tested them yet and therefore should not be merged, but I wanted to get this out there so people could try it.

technicalpickles commented:
I spent some time with this. Based on the current implementation, I don't think it has enough to be able to be used for adding NestCam support to HomeAssistant. In particular, HA needs to be able to get a still image.

Looking at the API reference (https://developers.nest.com/documentation/api-reference select Devices, filter out anything that isn't a Camera), it seems there's a `snapshot_url` attribute, that should be able download according to [How to Use Camera Objects](https://developers.nest.com/documentation/cloud/how-to-cameras-object/)

However, when I drop a IPython `embed` in the code, the device's `snapshot_url` looks like:

```python
'snapshot_url': {'snapshot_url_prefix': 'https://www.dropcam.com/api/wwn.get_snapshot',
  'snapshot_url_suffix': ''},
```

Which doesn't seem like it should be for a NestCam. I couldn't find any references for this, aside from googling and finding this snippet: https://github.com/tylerjaywood/pleasetakecareofmyplant/blob/master/time_lapse/screengrab.py

I was able to sort out the auth_token to use. And I thought I had the device id (`device.serial`). That gave me a method like:

```python
    @property
    def snapshot_url(self):
        return self._device['snapshot_url']['snapshot_url_prefix'] + self._device['snapshot_url']['snapshot_url_suffix'] + "/" + self.serial + "?auth=" + self._nest_api._session.auth.access_token

```

This results in technically a 200 response code when fetching it, but looking at the response gives:

```
{"status": 400, "items": [], "status_description": "bad-request", "status_detail": "Error retrieving snapshot for device XXXXXXXXXXXXXXXXXX"}.
```

I searched for other uses of this API call, and found [this pastebin](http://pastebin.com/jEiA4ZEC) with snapshot_urls. The thing being used as a device id didn't look like a serial.

tikismoke commented:
@jkoelker @technicalpickles Have you try playing with the nest home simulator to test support in case you do not owned ones?

technicalpickles commented:
@tikismoke I haven't. I don't even know if this API is the same as what's on https://developers.nest.com since it uses username/password, rather than access token and either OAuth or PIN. I do know this library is used by home-assistant though, and it has support for the other devices.

technicalpickles commented:
I had a bit of a breakthrough, and was able to use the new REST API, including testing against the [Nest Home Simulator](https://developers.nest.com/documentation/cloud/home-simulator).

The bad news is that it doesn't support email/password login, and you have to create both a developer account (developers.nest.com) and create a Product if you want to use this. There are also quite a few changes to their API, so it's going to take some time to start using the new things, and making sure that the old way of doing things has some new way. That'd be ideal for backwards compatability, but might be difficult. At worst, I think it'd be good to add some placeholder methods to raise an UnimplementedError or similar to explain it's gone away.

Here's where this branch is now: I've made changes so the thermostat attributes are loaded (read-only, no writing) when used with a branch of home-assistant (to be pushed soon). I've also updated the camera support to use the new API, and I was able to start a camera for homeassistant to sometimes see an image 🎉 

To use this branch, you'd need to get an access token as described in https://developers.nest.com/documentation/cloud/rest-quick-guide and pass that to `Nest` as a keyword arg

technicalpickles commented:
> Here's where this branch is now: I've made changes so the thermostat attributes are loaded (read-only, no writing) when used with a branch of home-assistant (to be pushed soon)

The branch is https://github.com/technicalpickles/home-assistant/tree/nest-cam but it is very rough and I haven't made a PR yet.

technicalpickles commented:
This has progressed quite a bit, and complete enough to have a PR to homeassistant using it merged: https://github.com/home-assistant/home-assistant/issues/4292

Now, this requires registering on https://developers.nest.com/ to create an account, and then [registering a product](https://developers.nest.com/products). This only supports using the [PIN-based authorization](https://developers.nest.com/documentation/cloud/authorization-overview) so far. That means anytime you try to use the API, you'd need to re-authorize the PIN. This can be annoying, so there is (and was) a way to cache an access token gotten through the PIN workflow.



```python
client_id = 'copy from your product'
client_secret = 'copy from your product'
nest = nest.Nest(client_id=client_id, client_secret=client_secret, access_token_cache_file="nest.json")
if nest.access_token is None:
    print('Go to ' + nest.authorize_url + ' to authorize, then enter PIN below')
    pin = input("PIN: ")
    nest.request_token(pin)
```

After that, you can access `nest` more or less like it already was. The main caveat is that the Nest API is totally different than what this library was using, in terms of what data was exposed.

I suspect this library had origins in discovering a mobile app had an internal API powering it, so it exposes a ton of things. The new 'proper' API that Nest exposes for developers has renamed a lot of things, changed some semantics, and removed many others. When possible, I try to find the thing that was semantically the same in the new API. When there wasn't, I raise NotImplementedErrors. I also raise that when there's no equivilant data in the new API.

technicalpickles commented:
This has been in released homeassistant for a few days now. There are definitely still kinks to work out, but there's been at least one PR to homeassistant to add more functionality to this, and it's using another fork's branch which is going to make merging and releasing slightly more complicated.

I think the main thing left is to update the documentation, which I'll try to do in the next day or two.

jkoelker commented:
@technicalpickles Awesome work here. Sorry for being MIA this fall, work took priority. I've looked through this and think its the right way to go. I'm adding you as a collaborator to the [repo](http://felixge.de/2013/03/11/the-pull-request-hack.html). I see that homeassistant is already using this, so its probably ready. Let me know if you think we should merge it and release it to pypi or wait.

technicalpickles commented:
@jkoelker thanks!

It is really close to being ready to merging & releasing on pypi. There were just a few cases I wanted to understand a little better, to see if they can be added later without affecting API (ie just a minor release with no API changes):

- detect when a product doesn't have the permissions to use the thing you want: https://nestdevelopers.io/t/documentation-about-when-attributes-may-be-omitted-from-response/492/3 talks about it a little. I guess it'd raise an error if there's no key? That technically changes the API if you currently rely on the None returned by `get`
- detect if it's necessary to re-authorize because the product's permisisons changed https://developers.nest.com/documentation/cloud/client-version/ (linked above)
- detect if it's necessary to re-authorize because the access_token failed to authenticate. right now, it's super long lived, but I think it's probably a reasonable thing to do.

Also, fix linting which I broke at some point 😁

technicalpickles commented:
@jkoelker sorry it took me this long to get back to it, but I think this is ready for merge and release. Not sure how you want to handle this with you merging and releasing to pip, or I can merge and you release to pip at some point.

jkoelker commented:
Awesome, I'll merge it, going to put a little note in the readme on how to stay pinned to the 2.x code for peeps that need to until they update. For pypi releasing we just have to push to the `deploy-test` branch to get it on https://testpypi.python.org/pypi then to `deploy` to get it on pypi proper. This way if I ever go mia for a while it can still be updated by whoever has access to the repo. If you have a pypi username I can also add you as an owner there too so we have more redundancy.

jkoelker commented:
Its now live on [pypi](https://pypi.python.org/pypi/python-nest/3.0.5).

technicalpickles commented:
Awesome, thanks 😁 

> If you have a pypi username I can also add you as an owner there too so we have more redundancy

I am technicalpickles on there too.

jkoelker commented:
Totes, added you to pypi as well. \o/.
